### ISSUE TITLE
/targets and /new/targets performance

### ISSUE THREAD

**Issue by carlpett** [prometheus/prometheus#6881]
**What did you do?**
I have a server with ~7k active targets (in a single job, if that matters). Opening /targets (or /new/targets) has really poor performance.

**What did you expect to see?**
#6276 was mentioned in #6186 as improving the performance, so page rendering in a couple of seconds-ish? No hard expectations, really.

**What did you see instead? Under which circumstances?**
The Classic UI displays a blank page while downloading ~8 MB in 2.5s, then downloads js/css stuff, finishing all downloads after 3.44s. It then finishes rendering after a total of 3.94s.
The New UI makes a few small quick requests, then downloads a 1MB js bundle, after which it renders most of the UI and a spinner. It then downloads a ~3.5 MB target list, and then hangs. After a couple of seconds Firefox prompts if I want to stop the js execution. If I opt to "Wait" instead, it will ask again after a couple more seconds, which seems to repeat until I ask it to "Stop", at which point everything renders.

**Discussion**
The root of all this seems to be that it is just too much data returned. Is it complicated to page the returned data?

**Environment**

* System information:

(Running in GKE)
```
Linux 4.14.127+ x86_64
```

* Prometheus version:

```
prometheus, version 2.16.0 (branch: HEAD, revision: b90be6f32a33c03163d700e1452b54454ddce0ec)
  build user:       root@7ea0ae865f12
  build date:       20200213-23:50:02
  go version:       go1.13.8
```

**Comment by roidelapluie** [comment]
It would probably not be complicated but would probably be breaking unless we explicitely add a per_page param ; which when unset returns everything.

**Comment by brian-brazil** [comment]
Pagination is hard, and mainly pushes the problem onto the server but even more so. That's probably not a good choice here given the importance of Prometheus itself.

**Comment by carlpett** [comment]
I'm not stuck on pagination per se, if there are better options. What are the hard things, in this case? That the target list could be changing is one somewhat-obvious thing where there might be correctness issues (which may or may not be considered a blocker - even with a non-paginated display the data is changing, but you need to refresh to see it). I don't see how this would be jeopardizing the stability of the server, however?
The current situation actually seems like **more** of a problem to server stability. Generating those thousands of targets into the json response seems to add a rather significant CPU load while it happens.

As an alternative, perhaps we could skip actual pagination and just cap the number of returned targets per job to some number?

**Comment by roidelapluie** [comment]
The targets page is super important as it is the only place we see failed scrape reasons. So not showing everything is not an option.

**Comment by brian-brazil** [comment]
> What are the hard things, in this case?

The sorting and resulting n^2 performance on the Prometheus server side, before considering that the list is not static.

> The current situation actually seems like more of a problem to server stability.

No, generating it once is far cheaper than doing so multiple times in the backend.

**Comment by carlpett** [comment]
Sidenote: I feel like this discussion fairly quickly derailed into a quite unconstructive tone?

@roidelapluie *If* there would be pagination, or limits, or whatever better solution it is possible to come up with, it would have to combine with the "show unhealthy" option that is already in place. So say we would show 50 (completely arbitrary number for discussion) failing targets, are there often cases where there are more than 50 different causes that you want to investigate in parallel? I don't think I've seen more than two different causes at the same time, and either way I wouldn't be able to tackle all of them at once.
This might also suggest that there should maybe be a better way to troubleshoot failing scrapes, but that would be the topic of another discussion.

@brian-brazil I of course don't have any data on this, but my observations of people using Prometheus over the years would suggest that _very few_ people would go through all the targets for all the jobs, so you'd end up not paying that cost for the majority case. Is this not in line with your experiences? Especially for larger servers, where this problem occurs.

**Comment by roidelapluie** [comment]
> Sidenote: I feel like this discussion fairly quickly derailed into a quite unconstructive tone?
> 
> @roidelapluie _If_ there would be pagination, or limits, or whatever better solution it is possible to come up with, it would have to combine with the "show unhealthy" option that is already in place. So say we would show 50 (completely arbitrary number for discussion) failing targets, are there often cases where there are more than 50 different causes that you want to investigate in parallel?

If the 51th is failing for another reason, and the 51th is somehow the proxy to the other ones? We can always come with pretty bad scenarios that are more or less science fiction.

> @brian-brazil I of course don't have any data on this, but my observations of people using Prometheus over the years would suggest that _very few_ people would go through all the targets for all the jobs, so you'd end up not paying that cost for the majority case. Is this not in line with your experiences? Especially for larger servers, where this problem occurs.

I often open the targets page to do a quick ctrl+f scan (for the record I have 4300 targets)

**Comment by brian-brazil** [comment]
> are there often cases where there are more than 50 different causes that you want to investigate in parallel?

The typical case where one scrape config is handling many many different types of application. It'd be usual for there to be "normal" failures there, which you wouldn't want to hide whatever you were actually looking for.

> Is this not in line with your experiences? Especially for larger servers, where this problem occurs.

My argument is that any filtering/pagination is for the frontend to handle, and that trying to do it in the backend adds an unnecessary risk of server load on the backend.

**Comment by carlpett** [comment]
Okay, given that a backend fix is not feasible/appropriate, perhaps something can be done on the frontend to deal with this a bit more gracefully? Especially the new React UI seems to deal with large volumes poorly, with hanging the browser (while the old/classic UI is merely slow).

**Comment by boyskila** [comment]
Yeah, 7k items in the DOM it's pretty heavy. Other option, different than pagination is the "virtualized" list. However, this approach requires extending the API to allows getting targets on chunks and.... unfortunately it will cut of ```CTRL+F``` searching.

**Comment by roidelapluie** [comment]
Maybe we can implement "lazy display" and add our own search text box ?

**Comment by carlpett** [comment]
It's been quite a while since I did any serious frontend development, but wouldn't it be possible to store the target list in memory without adding them to the DOM? Then as the user changes pages render them lazily (and optionally, if feasible, removing the previous page from the DOM).

I also realized that this discussion of course also applies to the service discovery page.

**Comment by boyskila** [comment]
> but wouldn't it be possible to store the target list in memory without adding them to the DOM

Yeah, this is what the virtualized list is.

**Comment by roidelapluie** [comment]
> > but wouldn't it be possible to store the target list in memory without adding them to the DOM Then as the user changes pages render them lazily
> 
> Yeah, this is what the virtualized list is.

Hence a small search box might help?

**Comment by boyskila** [comment]
> Hence a small search box might help?

Yeah, I guess this would be the only way to search for a target.

**Comment by boyskila** [comment]
I'm going to prepare some POC with the virtual list, I just need some data to test. Can you please guys send me the JSON with the data from the response attached as a file here?

**Comment by carlpett** [comment]
>>but wouldn't it be possible to store the target list in memory without adding them to the DOM Then as the user changes pages render them lazily
>Yeah, this is what the virtualized list is.

@boyskila Ok, then I misunderstood this bit?

>However, this approach requires extending the API to allows getting targets on chunks

We would still be fetching the full json output **once**, right? As mentioned by Brian above, incorporating any form of subset fetching from the API is out of scope.

I can share output in a couple of hours.

**Comment by boyskila** [comment]
> We would still be fetching the full json output **once**, right

Actually the virtual list approach doesn't care whether the data is partially coming, so yeah fetching it at once is fine.

**Comment by carlpett** [comment]
Sorry, of course forgot to share the file. We're at 16.6k targets now, so a bit larger than when I initially opened the ticket.
File has been renamed .txt since Github doesn't accept .json files :man_shrugging: 

[target.txt](https://github.com/prometheus/prometheus/files/4292839/target.txt)

**Comment by roidelapluie** [comment]
cc @juliusv did you see this?

**Comment by juliusv** [comment]
Yes, sorry! It's definitely a React render speed issue, similar to the one we have with https://github.com/prometheus/prometheus/issues/8548. But this one here is probably even more urgent, as thousands of targets are more common than thousands of alerts. There's probably no way around doing pagination and/or virtualized lists (although those seem to get more complicated in code when you don't know the height of each item in advance), at least if we keep those pages rendered by React. Planning to take a deeper look into this soon!

**Comment by roidelapluie** [comment]
@juliusv could we collapse by default when there is a huge list of targets?

**Comment by juliusv** [comment]
@roidelapluie That's an option too. It wouldn't help if someone then wanted to expand everything of course, but it would be a good stop-gap to make that page work again for people in the meantime.

**Comment by Nexucis** [comment]
Here we could have a different solution than the one proposed in #8548 (aka scrolling loading).

* Perhaps the UI should display the target list as closed by default
* But also what does it mean, it also required a lazy loading / rendering. Which means that the render of targets for the same job is not done until the user is clicking on `show more`.
* Like that you are able to load all target, but actually you don't generate the rendering for each of them.
It will help even if the user is clicking on every button to display all target, because the rendering of the different targets are done job by job, so it will reduce the CPU usage. 

The corner case is if you have one job with thousand and thousand targets. Then if you want to be sure it won't crash the browser, then you need a pagination / scrolling system.

* Also we could include a search bar as well (I think actually to have certain consistency between the different page, it would be interesting to have the search bar everywhere)

WDYT?

**Comment by Nexucis** [comment]
and perhaps someone could add the label `component/react-ui` since it's more an UI issue.

**Comment by juliusv** [comment]
@Nexucis Added the label!

> But also what does it mean, it also required a lazy loading / rendering.

I think we should only do the rendering lazily for now, not the loading (because the rendering seems to be taking most of the time, and adding lazy-loading may be hard for this). But I'm not sure if that would fully help on the level of a job / scrape pool, because all the targets could be in a single job (like in the original comment). So there may still have to be either pagination or virtual windowing (possibly with a search box).

**Comment by Nexucis** [comment]
yeah I think a pagination would be simple actually. Like that we will be sure it will limite the number of element to render.
Still I'm agree we should tackle first the lazy rendering and then (in an another PR) the pagination.

I think it would also be nice to have the same way to load element (loading scrolling vs pagination).

**Comment by juliusv** [comment]
@Nexucis I think normally we should just show everything (keeps the friction to look for something low), but only collapse individual scrape pools if there are too many targets in them, or if the total number of targets on the page (via too many scrape pools) would be very large. But at that point, it may make sense to only do the pagination and no closing-by-default?

**Comment by roidelapluie** [comment]
The more targets you have, the more likely they are changing a lot. As the API is not "transactional", pages are hard. So from the API I think we should still only  do one API call.

**Comment by juliusv** [comment]
@roidelapluie Yes, exactly.

**Comment by Nexucis** [comment]
oh from my point of view, the pagination is only on the UI side here (on this particular topic). So no pagination on the API side.

So basically the UI is downloading all the target, but then it creates a virtual pagination based on limit in the UI itself so it will avoid to display all targets at the same time and avoid to crash the browser because there are too many targets to display.

**Comment by Nexucis** [comment]
sorry for the bad explanation :/

**Comment by Nexucis** [comment]
I'm building a sort of "smart" search bar. Once it will be ready, I think we won't be so far to solve these issues

**Comment by Nexucis** [comment]
Just giving an update on this issue : I finished a first implementation of the "smart" search bar and it has decent performance when searching through 10k targets.
Just need to publish the package now. In case someone is interested, the code is here: https://github.com/Nexucis/kvsearch

So after the holidays, I will focus on how we can have a lazy loading of the targets (like with an infinite scrolling).
I'm thinking of deeply review the `Target` page (in term of code, not necessarly in term of view).

Hopefully soon, I should be able to produce a draft PR to fix this issue.

**Comment by Nexucis** [comment]
PR #10103 will fix this issue.
